### DISCUSSION TITLE
Invariant: expected pageData to be a string, got undefined errors on App Router pages

### DISCUSSION THREAD

**Discussion by jcole356** [vercel/next.js#68174] [Help]
### Summary

With international traffic enabled we occasionally see high levels of traffic that we suspect to be security scans.  During these periods we eventually start seeing `Invariant: expected pageData to be a string, got undefined` errors.  Once one of our instances is impacted by this error, that instance never recovers and the cache for whatever component was being accessed seemed to remain tainted.  We have to restart the instance to recover.  Otherwise we start seeing `_error` pages.

The most relevant part of our deployment stack is as follows:

2 edge servers (corporate mandate)
Load balancer
AWS Fargate hosted NGINX proxy
AWS Fargate hosted Next.js server

Our stack traces in New Relic are only detailed enough to tell us where in the Next.js source code the error is being thrown.  There is no indication where in our code this error occurs.  We have been unable to reproduce on our own.

Our current suspicion is there is something malicious in some of the requests, possibly a header that is being abused, but we have not been able to find it to this point in the logs.

**Comment by shawngustaw** [comment]
@jcole356 any chance you figured out what the issue was? We're currently running into a similar scenario

**Reply by jcole356** [reply]
> @jcole356 any chance you figured out what the issue was? We're currently running into a similar scenario

@shawngustaw we have not seen this error for a while.  Unfortunately we were never able to completely lock down the exact cause because the issue was intermittent.  The following are some of the things we did:

[Implement a custom caching strategy with Redis](https://nextjs.org/docs/app/building-your-application/deploying#configuring-caching)

This does not get enough attention in the docs IMO.  If you have a multi instance environment, Next.js's in memory cache is not going to work out of the box.

[Disable prefetching for Next.js links](https://nextjs.org/docs/app/building-your-application/deploying#configuring-caching)

I combed through the Next.js source code surrounding the `Invariant` error we were seeing.  Prefetching seemed to be referenced in the code path.  We were not getting a notable return from enabling, so we disabled suspecting it might not be playing nice with our self hosted environment and CDN.

Block all non essential headers consumed by Vercel. ([see relevant blog](https://zhero-web-sec.github.io/research-and-things/nextjs-and-cache-poisoning-a-quest-for-the-black-hole))

This one is trickier, but Vercel uses custom headers under the hood for their deployments.  In a self hosted environment, some of these are unnecessary, you can block them.
